cli/src/config.rs

```diff
@@ -162,27 +163,65 @@ impl AsMut<StackedConfig> for RawConfig {
     }
 }
 
+#[derive(Copy, Clone, Debug)]
+struct DeprecatedLocation {
+    old: &'static str,
+    new: &'static str,
+}
+#[derive(Clone, Debug)]
+enum ConfigPathState {
+    New,
+    Exists,
+    Deprecated(DeprecatedLocation),
+}
+
+/// A ConfigPath can be in one of two states:
+///
+/// - exists(): a config file exists at the path
+/// - !exists(): a config file doesn't exist here, but a new file _can_ be
+///   created at this path
+///
+/// Additionally, if it exists, it may be deprecated. In this case, jj will warn
+/// when ConfigEnv::warn_for_deprecated_paths is called.
 #[derive(Clone, Debug)]
-enum ConfigPath {
-    /// Existing config file path.
-    Existing(PathBuf),
-    /// Could not find any config file, but a new file can be created at the
-    /// specified location.
-    New(PathBuf),
+pub struct ConfigPath {
+    path: PathBuf,
+    state: ConfigPathState,
 }
 
 impl ConfigPath {
     fn new(path: PathBuf) -> Self {
-        if path.exists() {
-            ConfigPath::Existing(path)
-        } else {
-            ConfigPath::New(path)
+        use ConfigPathState::*;
+        ConfigPath {
+            state: if path.exists() { Exists } else { New },
+            path,
         }
```

Comment:
nit: is there a reason these aren't provided in the order they're defined?

```suggestion
        ConfigPath {
            path,
            state: if path.exists() { Exists } else { New },
        }
```

Reply:
Yes, because the `path` line moves the path that the `state` line refers to.